### ISSUE TITLE
No taxcalc package for Python 3.6

### ISSUE THREAD

**Issue by martinholmer** [open-source-economics/Tax-Calculator#1339]
For some time now the taxcalc package has been made for Python 2.7, Python 3.4, and Python 3.5.
You can see that from the following command:
```
$ conda search -c ospc taxcalc* | grep 0.8
                             0.8.0               np111py35_0  ospc            
                             0.8.0               np111py34_0  ospc            
                             0.8.0               np111py27_0  ospc            
                             0.8.1               np111py27_0  ospc            
                             0.8.1               np111py35_0  ospc            
                             0.8.1               np111py34_0  ospc            
                             0.8.2               np111py34_0  ospc            
                             0.8.2               np111py27_0  ospc            
                             0.8.2               np111py35_0  ospc            
```
However, Anaconda distributions of Python 3.4 and Python 3.5 are not available on the Continuum Analytics [download page](https://www.continuum.io/downloads); only Python 3.6 and Python 2.7 are available there.

So it seems as if we can recommend to CLI users only Python 2.7 because we have no conda taxcalc package for Python 3.6.  Right?  If not correct, what am I missing?

Why are we still making the out-of-date Python 3 packages?  Shouldn't we be making taxcalc packages only for the versions of Anaconda Python available from Continuum Analytics?

Answering those questions by saying we need to serve people using older versions of Python 3 is not convincing because if that were really the goal we would have been making packages for 3.0, 3.1, 3.2 and 3.3 as well.

@MattHJensen @feenberg @Amy-Xu @andersonfrailey @GoFroggyRun @codykallen 
@PeterDSteinberg @brittainhard

**Comment by PeterDSteinberg** [comment]
The build is currently dying on Python 3.6 and has never succeed on Python 3.6.  There is a dependency issue that prevents successful conda build.  This issue can be a reminder to check into the package specs for taxcalc as they related to Python 3.6 and see what is the conflict.

**Comment by martinholmer** [comment]
@PeterDSteinberg said:
> The build is currently dying on Python 3.6 and has never succeed on Python 3.6. There is a dependency issue that prevents successful conda build.

Can you tell us more about the "dependency issue"?

@MattHJensen @feenberg @Amy-Xu @andersonfrailey @GoFroggyRun @codykallen @brittainhard

**Comment by PeterDSteinberg** [comment]
@martinholmer This is what I just got in Jenkins:
```
The following specifications were found to be in conflict:
  - pandas <=0.18.0 -> numpy 1.9* -> python 2.7* -> openssl 1.0.1*
  - python 3.6*
```

**Comment by feenberg** [comment]
On Wed, 3 May 2017, Peter Steinberg wrote:

> 
> @martinholmer This is what I just got in Jenkins:
> 
> The following specifications were found to be in conflict:
>   - pandas <=0.18.0 -> numpy 1.9* -> python 2.7* -> openssl 1.0.1*
>   - python 3.6*
> 
> —

Do we gain anything from the incompatible code that creates these 
restrictions? I can't judge the dificulty of writing more compatible code, 
but I am very skeptical of existence of serious benefits from code that 
requires particular versions of pandas or numpy or (???) openssl. The 
instructions for writing compatible Python code:

    http://python-future.org/compatible_idioms.html

are long, but surely mostly about things that should never come up in the 
tax calculator. Are we quite sure that the requirements of the calculator 
will not make it difficult to install unrelated software with different 
requirements? Here at NBER we are plagued by R packages requiring 
different and incompatible libraries. While my system administrator is 
willing to put in the effort to allow the conflicting packages to 
co-exist, I doubt if many others are.

dan feenberg

**Comment by martinholmer** [comment]
A week or so ago there was some discussion in issue #1339 about why there is no taxcalc package for Python 3.6, which is the only version of Python offered on the Continuum Analytics Anaconda download page except for the legacy 2.7 version.

As part of that conversation, @PeterDSteinberg said:
> The following specifications were found to be in conflict:
>  - pandas <=0.18.0 -> numpy 1.9* -> python 2.7* -> openssl 1.0.1*
>  - python 3.6*

I'm not certain about how to decode that, but it seems to involve the (then) requirement that Tax-Calculator use a version of panda no higher than 0.18.0.

Another part of that conversation was what @feenberg said:
> Do we gain anything from the incompatible code that creates these restrictions?

It seemed that Dan thought the reason we could not create a taxcalc package for Python 3.6 was because we had written Python code in a non-portable way.

In order to better understand these issues, I tried to build a taxcalc package for Python 3.6 on my local computer.  I had no problems doing that.  And I used that package to run the tc CLI with `puf.csv` input and produced the same results and graphs as I have been producing under Python 2.7.

This suggests that there is no problem with the generality of our Python source code.  And it also suggests that the root of the 3.6 problem was the (old) restrictive requirement that Tax-Calculator use an old version of pandas, which was changed for other reasons by the merge of pull request #1351.

Given these results, **I suggest that we quit making a taxcalc package for Python 3.4 and start making a taxcalc package for Python 3.6**.  Any reactions or comments to that proposal?

@MattHJensen @Amy-Xu @andersonfrailey @GoFroggyRun @codykallen @zrisher @brittainhard

**Comment by MattHJensen** [comment]
That makes sense to me, too. 

>

**Comment by martinholmer** [comment]
@MattHJensen said in issue #1339:
> That [dropping 3.4 and adding 3.6] makes sense to me, too.

OK.  What edits do I make in the Tax-Calculator repository to drop the 3.4 tests on GitHub and add the 3.6 tests on GitHub?  Or is all that controlled by somebody else in some other code base?

**Comment by martinholmer** [comment]
@MattHJensen,  I answered my own question about how to drop testing in 3.4 environments and add testing in 3.6 environments.  See the changes in pull request #1361.

**Comment by martinholmer** [comment]
Issue #1339 has been resolved by pull request #1361.
We are now testing and building a taxcalc package for Python 3.6.

@MattHJensen @feenberg @Amy-Xu @andersonfrailey @GoFroggyRun @codykallen @zrisher 
@PeterDSteinberg @brittainhard
